BUG: define "uint-alignment", fixes complex64 alignment

ahaldane opened the pull request:
This PR fixes alignment checks along the lines of #5365 and [this comment](https://github.com/numpy/numpy/pull/5365#issuecomment-66690535).

This fixes the bug that `complex64` (and `complex128`) has incorrect alignment, and should be 4 instead of 8 on most systems. This has caused further bugs such as unnecessary buffer allocation, (previously) bus errors, and more work for people adding new architectures.

The problem is that numpy really needs two different types of alignment check: 
- "true alignment" is the alignment required by the architecture for safe/fast access (used by ufuncs/casting/copyswap)
- "uint alignment" is the alignment of an equal-sized uint, required by the strided-copy functions. which get a speedup relative to `memcpy(dst, src, N)` by doing `*((uintN*)dst) = *((uintN*)src)`.

See discussion in #5365, #5316, #3816, #5656, in the docs committed here, plus the bit at the end of the [original notes](https://gist.github.com/ahaldane/6f57412470c3d8f12c20) (which the docs are based on).

This PR splits up the two types of alignment checks into two sets of functions, the first for uint alignment, the second for "true" alignment.

ahaldane committed:
TST: Test complex64 alignment

ahaldane committed:
DOC: Document how memory alignment works as of 1.14

charris commented on numpy/core/src/multiarray/lowlevel_strided_loops.c.src:
```diff
@@ -1382,7 +1372,7 @@ mapiter_trivial_@name@(PyArrayObject *self, PyArrayObject *ind,
     npy_intp itersize;
 
-    int is_aligned = PyArray_ISALIGNED(self) && PyArray_ISALIGNED(result);
+    int is_aligned = IsUintAligned(self) && IsUintAligned(result);
```
`IsUintAligned` needs a declaration.

charris commented:
Looks like `array_assign.h` needs to be included in a few files. @juliantaylor Could you take a look?

ahaldane commented:
Right now I'm looking into the python3 assertion failures which seem more serious.

ahaldane commented:
I'll have to finish later. But this is what fails:

    >>> np.copyto(np.array([0]), np.array([1]), where=np.array([False], dtype=bool))

My current guess is that there is a bug in `_strided_masked_wrapper_transfer_function`: It first tries looks for the next-non masked value using `npy_memchr`, but there is none. Then `subloopsize` is 1, and `dst_stride` is -1, which causes dst to point to some invalid address. Then it attemps to copy values to that address, which trips the assertion in  `_aligned_strided_to_strided_size8_srcstride0`.

I think this only turned up now because I added `RELAXED_STRIDE_CHECKING` to `raw_array_is_aligned`.

juliantaylor commented:
is there anything different to my branch besides adding raw_*
would be easier to review without all the unnecessary variable renames

ahaldane commented:
@juliantaylor I renamed all the uintaligned for now, but I'd like to keep the change in the final version, since I think it will be confusing that 'aligned' does not actually mean aligned.

ahaldane committed:
BUG: _strided_masked_wrapper_transfer_function goes out of bounds

ahaldane commented:
OK, I think I fixed the bug in `_strided_masked_wrapper_transfer_function` (see last commit).

@juliantaylor It's largely the same, but I reorganizes the `IsAligned` and `raw_array_is_aligned` functions to avoid duplication, removed the special case for flexible arrays, and implemented the "uint aligned" case differently.

The last 4 (unnumbered) commits are additional bugfixes/cleanups.

ahaldane commented:
I added a few more changes I missed before (last commit). Nditer code needed "uint" alignment too.

Also, I got rid of `NPY_MAX_COPY_ALIGNMENT`, since I'm pretty sure it isn't needed any more, since the "uint" alignment checks make everything safe.

pitrou commented on doc/source/dev/alignment.rst:
```diff
@@ -0,0 +1,93 @@
+   determines offsets automatically. In that case, ``align=True`` pads the
+   structure so that each field is aligned in memory, sets ``dtype.alignment``
+   to be the largest of the field alignments, and sets ``dtype.itemsize`` to
+   the smallest posible multiple of this alignment. This is what C-structs
```
(it also sets `dtype.isalignedstruct`)

homu commented:
:umbrella: The latest upstream changes (presumably #7134) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7481) made this pull request unmergeable. Please resolve the merge conflicts.

ahaldane committed:
BUG: raw_array_is_aligned ignores NPY_RELAXED_STRIDES_CHECKING

ahaldane committed:
ENH: Implement methods for uint-alignment

Implements IsAligned, IsUintAligned, npy_uint_alignment

ahaldane commented:
Rebased and simplified.

In summary:
 * Fixes a few bugs in preparation for the other commits (first 2 commits, see [comment](https://github.com/numpy/numpy/pull/6377#issuecomment-143597287))
 * Implements functions to differentiate "uint alignment" and "true alignment" (see doc commit). The "uint alignment" computation is quite different (I think more literal) than in [Julian's branch](https://github.com/juliantaylor/numpy/commit/698205f874abea12552993864778e46d9df7f8c3).
 * Switch over all copy/transfer code paths to use the "uint alignment". This involved a lot of grepping/reading to catch all the uses of `_IsAligned`, `PyArray_ISALIGNED`, `raw_array_is_aligned` in many numpy files.

I've removed the massive variable renaming  I used to have of "aligned" -> "uintaligned" in the copy code paths. We just need to remember that in those files "aligned" refers to uint alignment, not true alignment.

ahaldane committed:
ENH: Fix complex64 alignment

ahaldane committed:
ENH: Make copy-code-paths check for uint-alignment

charris commented:
Does this have any back compatibility effects that would need a release note?

charris commented:
I think we will want to replace the CamelCase function names at some point, but that can be another PR.

ahaldane commented:
Yes, changing the alignment of complex64 is a behavior change that could affect users. For example, it will change the size of aligned structured types, like `np.dtype('c8,u1', align=True).itemsize` (which is now 12, like a C struct, instead of 16).

I'll add a note.

charris commented:
Does this affect pickles, npz files, etc.? I'm mostly concerned about data stored with the old alignment becoming invalid. Don't know if that is possible or not, I'm guessing that the stored offsets will take care of that and that if it were a problem cross platform compatibility would already be broken.

ahaldane commented:
That's an important thought! But I don't think it's a problem:

I don't think it affects npz file loading because those files store the dtype using the pep3118 data-syntax (`dt.descr`), which explicitly includes any padding bytes. So for instance, `np.dtype('c8,u1', align=True).descr` is `[('f0', '<c8'), ('f1', '|u1'), ('', '|V7')]` in current master, which explicitly includes 7 trailing padding bytes. (After this PR it will become `[('f0', '<c8'), ('f1', '|u1'), ('', '|V3')]`). When you `np.load` either of those, before and after this PR, you will get the padding bytes used at `np.save` time.

As a sidenote, `np.load` currently horribly screws up any such datatype with padding bytes anyway, see my old unfinished PRs #7798 and those related to #8100. So it is unlikely anyone has stored aligned structs in npz files. That motivates getting this alignment bug fixed *before* fixing #8100, since people may start storing padded structs after #8100 is fixed making it harder to change alignment bugs.

charris commented on numpy/core/src/common/array_assign.c:
```diff
@@ -84,14 +84,27 @@ broadcast_error: {
+                     char *data, npy_intp *strides, int alignment)
 {
     if (alignment > 1) {
         npy_intp align_check = (npy_intp)data;
```
Maybe `npy_uintp`. I'm pretty sure both work here for twos complement integers, but I needed to think about it.

charris replied:
Note that it is later cast to `npy_uintp` in `npy_is_aligned`.

pitrou replied:
Only bitwise operations are involved, so it should be exactly the same.

charris commented on numpy/core/src/common/array_assign.c:
```diff
@@ -84,14 +84,27 @@ broadcast_error: {
+            /* skip dim == 1 as it is not required to have stride 0 */
+            if (shape[i] > 1) {
+                /* if shape[i] == 1, the stride is never used */
+                align_check |= strides[i];
```
I think there is an assumption of multiples of powers of two, together with twos complement arithmetic, built in here. AFAICT, there are no false positive, but perhaps some false negatives. @juliantaylor @seberg Thoughts?

pitrou replied:
I don't see any arithmetic here, only bitwise OR operations. Whether the integer is considered signed or unsigned doesn't come into play.

seberg replied:
Well the bitwise or is used together with arrhythmic stuff later since it is used in `npy_is_aligned`. Did not think about it long, but I think you are right Chuck. I guess `npy_is_aligned` might explain it a bit more?

seberg replied:
Not sure if twos complement has anything to do with it though.

pitrou replied:
"arrythmic"? :-)

More seriously, I don't know what you are bothering about. First `data` is cast to `npy_intp`(this is an exact bitcast), some bitwise operations are done on the signed integer (they are agnostic wrt signedness, they just operate on individual bits), then the result is cast back to `void*` (this is an exact bitcast), then it is finally cast to `npy_uintp`.

There is zero reason to believe that the intermediate cast to `npy_intp` is any potential cause for trouble.

eric-wieser replied:
Casting unsigned integers to signed integers is implementation-defined

charris replied:
> Not sure if twos complement has anything to do with it though.

For example, ones complement turns even into odd when cast to unsigned.

pitrou replied:
You mean when ORing `strides[i]` with `align_check`? I would be surprised if the compiler produced anything else than a pure bitwise OR here. You may want to enable conversion warnings to see if it says anything.

pitrou replied:
Anyway, I don't even know why I'm commenting on this PR. I have other things to do.

charris replied:
> I guess npy_is_aligned might explain it a bit more?

That function is pretty straight forward, but I think what happens with the or's is that we are guaranteed the largest power of two that divides all the possible addresses. I think what we would really like is the GCD of all the used addresses, which reduces to the GCD of the address and the stride, and then check if that is divisible by the alignment. But that is more complicated.

charris replied:
> Again, these are pure bitwise casts here.

Yes, I know. But one does lose some information with two complement. Note, for instance, on a three bit machine, 3 (011) and -3 (101) are both divisible by 3 but 111 is not.

seberg replied:
Well, if this errs too much on the safe side (no idea if it does). I guess placing  the `npy_is_aligned` check into the loop here instead should work and probably not be a lot of overhead. (Plus this is the flags calculation, right? so it doesn't happen very often anyway). Or am I missing something? Plus it might be a bit less mind boggling.

charris replied:
I was planning to sleep on it :) I think when all is done, it will be OK for its intended application, but I want to understand it...

ahaldane replied:
I think you're right that this assumes a twos-complement representation of signed integers, even though the C standard explicitly allows others like ones-complement. I'm not sure we care, because in practice all modern computers use two's complement, and because all this pointer bit-twiddling is C-implementation-dependent anyway.

Here's my understanding: We are trying to compute whether `(data + n*stride)%alignment == 0` for all integers `n`. First, assuming alignment is a power of two, and assuming twos-complement representation, simplify the `%alignment == 0` to checking if the lowest log2(alignment) bits are 0. Next, use the fact that in twos complement representation, if `x` low order bits of `data` and `stride` are 0, then the `x` lower bits of `data + n*stride` must also be 0 for all `n` because of how binary addition works. Neither simplification is valid for negative values in ones-complement representation.

Casting the intp `stride` to uintp` and using uintp would avoid the signed twos-complement assumption, and the algorithm still works.

However, the representation/cast of `data` from pointer to uint is implementation defined too, so we are violating the standard anyway anytime we use bitwise ops on pointer values. (only arithmetic is allowed, I think). I don't see any alternative though, as far as I see there is no totally standard-compliant way to test alignment.

charris commented on numpy/core/src/multiarray/lowlevel_strided_loops.c.src:
```diff
@@ -1385,7 +1375,7 @@ mapiter_trivial_@name@(PyArrayObject *self, PyArrayObject *ind,
 
     npy_intp itersize;
 
-    int is_aligned = PyArray_ISALIGNED(self) && PyArray_ISALIGNED(result);
```
How do `PyArray_ISALIGNED` and `IsUintAligned` differ?

charris commented on doc/source/dev/alignment.rst:
```diff
@@ -0,0 +1,94 @@
+datatype is implemented as ``struct { float real, imag; }``. This has "true"
+alignment of 4 and "uint" alignment of 8 (equal to the true alignment of
+``uint64``).
+
```
So there are some structures that have no `uint` alignment?

ahaldane committed:
MAINT: remove unneeded test in npy_is_aligned

eric-wieser commented on numpy/core/src/common/array_assign.h:
```diff
@@ -87,10 +87,25 @@ broadcast_strides(int ndim, npy_intp *shape,
 /*
  * Checks whether a data pointer + set of strides refers to a raw
- * array which is fully aligned data.
+ * array whose elements are all aligned to a given alignment.
```
Can you add the remark about `alignment` being a power of two here?

ahaldane commented on doc/source/dev/alignment.rst:
```diff
@@ -0,0 +1,96 @@
+"Uint" alignment depends on the size of a datatype. It is defined to be the
+"True alignment" of the uint of equivalent size to the datatype of interest,
+or undefined/unaligned if the size is not a power of two. If the itemsize of an
+array is not a power of two the array can never be uint aligned.
```
I'll have to rephrase, that's an inaccurate sentence. It also doesn't describe the fact that 16 bytes types have a uint alignment of 8, because that's what the copy code wants. Maybe:

```
"Uint" alignment depends on the size of a datatype. It is defined to be the
"True alignment" of the uint used in numpy's copy-code to copy the datatype,
or undefined/unaligned if there is no equivalent uint. Currently numpy uses 
uint8, uint16, uint32, uint64 and uint64 to copy data of size 1,2,4,8,16 bytes 
respectively, and all other sized datatypes cannot be uint aligned.
```

charris commented:
Thanks Allan.
